fix: misc fixes to swift script on macos, incl URL inspection

iloveitaly opened the pull request:
- Missing clean statement in makefile
- Extract url from safari and chrome-like browsers

The new macos window watcher is working great so far! Will do some more testing over the next couple weeks.

iloveitaly commented:
@ErikBjare @modderme123  mind taking a look at this one?

modderme123 commented:
Looks good, although if you are removing the idle detection loop you might be able to remove the system idle time function as well

iloveitaly commented:
Nice tip! Just removed in the latest commit.

@modderme123 do you recall what this comment is referring to?

>TODO: set proper pulsetime

What's not right about the current pulsetime setup?

modderme123 commented:
Here's how I understand pulsetime is used: https://docs.activitywatch.net/en/latest/_modules/aw_transform/heartbeats.html#heartbeat_reduce

This is very much intended for a polling system so that it doesn't ever need to diff anything, and can just send the latest value, and it will get merged if the pulsetime is long enough.

Because we intentionally don't poll, pulsetime doesn't make as much sense for us. On a new event, we just want to make sure the previous event duration is extended until the start of our new event, and the new event is created ASAP.

To extend the previous duration, we send the previous event again (except at the current time), with a pulsetime long enough that it gets merged. Then it doesn't really matter what the new event pulsetime is, because it shouldn't get merged (it necessarily has a different value, because our notifier was triggered).

I assume the original comment added by @ErikBjare was referring to wanting to switch partially over to a hyrbrid polling model like the chrome extension does (use a notifier for precise updates and then send them on a timeout, which seems like it accidentally loses resolution if you rapidly switch between apps)...

modderme123 commented on aw_watcher_window/macos.swift:
```diff
@@ -85,13 +72,45 @@ struct Heartbeat: Codable {
+
+func logTimestamp() -> String {
+  let now = Date()
+  let formatter = DateFormatter()
```
Maybe extract this out so that you don't create a formatter every time you want to format a date?

iloveitaly replied:
Good call, I'll make that change.

iloveitaly commented:
@modderme123 thanks for the comment, very helpful! 

@ErikBjare this is ready for review and is working great for me locally. The main changes here are:

* improved the logging with levels & timestamp
* extract url information from safari
* refactored logic and added lots more comments. I _think_ I broke something when I refactored, which is what caused the conversation in discord.

modderme123 commented on aw_watcher_window/macos.swift:
```diff
@@ -225,13 +298,26 @@ class MainThing {
+      let frontWindow = safariObject.windows!()[0]
+      let activeTab = frontWindow.currentTab!
+
+      // Safari doesn't have an incognito mode, we cannot hide the url
```
This isn't true, we just need to figure out how to tell if the current window is in private browsing mode

modderme123 replied:
For example, using menus: https://alexwlchan.net/2021/04/detect-private-browsing/

iloveitaly replied:
Sorry, comment was bad. Here's what I meant to say, which I _think_ is true:

```suggestion
      // Safari doesn't allow incognito mode to be inspected, so we do not know if we should hide the url
```

iloveitaly replied:
good call, added an issue for this here https://github.com/ActivityWatch/aw-watcher-window/issues/71

iloveitaly commented on aw_watcher_window/macos.swift:
```diff
@@ -158,54 +184,60 @@ func createBucket() {
 
-  // Then, send heartbeat starting new event
-  let payload_new = heartbeat
+  // if you resize a window a ton of events (subsecond) will be fired
```
Ah, this is definitely better. We can compare app & title instead—I'll do this.

iloveitaly commented on aw_watcher_window/macos.swift:
```diff
@@ -158,54 +184,60 @@ func createBucket() {
+        let refreshedOldHeartbeat = Heartbeat(
+          // it is important to refresh the hearbeat using the timestamp where the user stopped working on the previous application
+          // TODO unsure if this is necessary: does the timestamp of the old vs new event need to be distinct
+          timestamp: heartbeat.timestamp - 1,
```
What I wasn't certain about here if we can send two events with the same timestamp (i.e. the old, refreshed heartbeat) and the new heartbeat. Sounds like you didn't have any issues with that?

(this was me being lazy and not testng)

modderme123 replied:
Oh I understand what this is doing now. Are the units ms or s? This seems reasonable if it is only ms

ErikBjare replied:
>Do we know if we can send two events with the same timestamp without causing any issues?

It *shouldn't* cause issues, but I've guarded for similar situations in other places by subtracting/adding 1ms to events.

iloveitaly replied:
Got it, I'll drop it to a millisecond subtraction and call it a day. I think all comments are addressed—good to merge!

ErikBjare commented:
@iloveitaly Waiting for you to address @modderme123's comments :)

iloveitaly commented on aw_watcher_window/macos.swift:
```diff
@@ -342,6 +390,7 @@ class MainThing {
   }
 }
 
+// TODO I believe this is handled by the python wrapper so it isn't needed here
```
Yeah, I don't think it's a massive deal either way just additional overhead to make sure they are hitting the right server APIs.

iloveitaly commented:
@ErikBjare @modderme123 comments addressed!

ErikBjare commented:
@iloveitaly Before merging this, does this resolve https://github.com/ActivityWatch/activitywatch/issues/787?

iloveitaly commented:
@ErikBjare I _think_ so, although I wasn't running into the identical issue he experienced. My window bucket was empty, but I had no issues with accessibility permissions. I've been running this for a couple weeks at this point and my window activity is back to normal on macos.

ErikBjare commented:
Alright, I tested it and it worked for both Chrome and Safari.

However, I decided to dig a bit deeper, and removed the Automation permissions for my terminal (Alacritty). After that, titles and URLs were blank again.

At minimum, it feels like we should at least get a window title even without Automation permissions, right?

<img width="780" alt="Screenshot 2022-10-26 at 16 42 17" src="https://user-images.githubusercontent.com/1405370/198057411-01513cb0-dd05-44be-948d-bb1ab157f4b3.png">

(took the screenshot after I enabled again, at which point it started working again)

ErikBjare commented on aw_watcher_window/macos.swift:
```diff
@@ -225,13 +299,26 @@ class MainThing {
         data.url = activeTab.URL
         if let title = activeTab.title { data.title = title }
```
This check needs to not only `if let title = activeTab.title { ... }`, but also if the string is empty, to prevent that bad permissions to lead to empty titles (I assume).

Is there a reason to do title assign here at all? Won't it just be the same?

iloveitaly replied:
Not sure if there is a reason we don't just use the original title. Make a change locally to log if there is a diff and will report back.

iloveitaly replied:
The title is different. Reason is the tab title is different than the window title in many cases. I've updated the code to conditionally use the tab title if it is not blank. Running this locally for a week and then will report back.

ErikBjare replied:
Can you provide a short example of a differing title? Curious what kind of differences there are

ErikBjare commented on aw_watcher_window/macos.swift:
```diff
@@ -225,13 +299,26 @@ class MainThing {
+
+      // Safari doesn't allow incognito mode to be inspected, so we do not know if we should hide the url
+      data.url = activeTab.URL
+      if let title = activeTab.name { data.title = title }
```
Same as above comment.

ErikBjare reviewed (commented):
Trying to get to the bottom of this browser issue.

ErikBjare commented:
Not at my Macbook, so can't test it myself, but going to eagerly assume this is good to merge.

Thanks again for the contributions @iloveitaly! Sorry for the slow merge :)
